odoo/modules/registry.py

```diff
@@ -30,6 +30,9 @@
 _logger = logging.getLogger(__name__)
 _schema = logging.getLogger('odoo.schema')
 
+# total dict capacity (as a power of 2)
+_CACHE_BASE = 14 # 16384
+_CACHE_SIZE = int((2**_CACHE_BASE - 64) * 2 / 3)
```

Comment:
Why do you subtract 64?  And please mention the reason for the 2/3 factor in the comment above.  The comment may be concise, like: "the dict allocates 2**N slots but uses maximum 2/3 of them (load factor); using this formula avoids wasting memory space". At least a developer finding this will quickly figure out the reason without having to dig code history.

Reply:
Not sure, but I since `dk_entries[]` is an array, I expect it adds item at `largest_used_index + 1` when `set` and marks item empty when `del`.  LRU popping cannot decrease the `largest_used_index` without resizing. 

CACHE_BASE = 1
CACHE_SIZE = 1
[a,,] -set-> [a,b] -del-> [empty, b] -set-> -**resize**-> [b, c] -del-> [empty, c] -set-> -**resize**-> [c, d] -del-> [empty, d]


CACHE_BASE = 1
CACHE_SIZE = 2
[a, b] -set-> -resize->-**resize**-> [a,b,c, ,] -del-> [empty, b, c, ,] -set-> [empty, b, c, d,] -del-> [empty, empty, c, d]
it cost more RAM but less **resize**

If `every __setitem__ checks the limit`, the 64 means after 64 dict set_then_pop, the dict will be resized once. So considering the size of the LRU cache, the 64 could be a bigger one 🤔

Reply:
I read the dictobject.c a little bit.
From the perspective of PyDictObject memory usage, I think `_CACHE_SIZE = int((2**_CACHE_BASE) * 2 / 3)` is already good enough.
However `dk_entries[]` only saves some pointers and (its values are pointers of PyObject). Comparing with the real data saved for LRU cache, I think the memory overhead of these unused dk_entries can actually be ignored. 🤔 

From what I understand, the real size used for a PyDictObject depends on the `dictresize` and its `GROWTH_RATE`
https://github.com/python/cpython/blob/6a80664ef1200008d5beb1584e03d779ef30cc58/Objects/dictobject.c#L1194
The GROWTH_RATE is `#define GROWTH_RATE(d) ((d)->ma_used*3)`
so after resizing, the new size of `dk_indices[]` should be `1 << ceil(log(LRU_cache_size * 3, 2))`
so the new size of `dk_entries[]` should be `(1 << ceil(log(LRU_cache_size * 3, 2))) * 2 // 3`


If we choose a good LRU_cache_size, after resizing, we will use around 50% of the `dk_entries`
and the LRU cache will resize after caching around LRU_cache_size new items

In the test script appended, the `LRU_cache_size` is set to 10922.
the size of `dk_indices[]` is 32768
the size of `dk_entries[]` is 21845
And you can see the dict is resized after del+insert 10924 items

```py
import time
import functools
import math

_CACHE_BASE = 14 # 16384
_CACHE_SIZE = int((2**_CACHE_BASE) * 2 / 3)

cache_size = _CACHE_SIZE
print('cache size: ' + str(cache_size))
# cache size: 10922
indices_size = 1 << math.ceil(math.log(cache_size * 3, 2))
print('indices size: ' + str(indices_size))
# indices size: 32768
entries_size = (indices_size << 1) // 3
print('entries size: ' + str(entries_size))
# entries size: 21845

# test dict
def del_insert_pydict(pydict, i):
    # del an item and insert a new one
    pydict_size = len(pydict)
    s_time = time.time()
    # old item is deleted before the new one is added
    # https://github.com/python/cpython/blob/2405929c35239580aa0344f0e8fd31b87f9ccfe9/Lib/functools.py#L608
    del pydict[i]
    pydict[i + pydict_size] = 'a'
    e_time = time.time()
    return e_time - s_time

# get the resize time
pydict = {i: 'a' for i in range(cache_size)}
t_resize = del_insert_pydict(pydict, 0)
print('\nthe estimated resize time is:')
print(str(t_resize) + 's')
# 7.390975952148438e-05s

print('\nthe dict is resized after del+insert the following number of items:')
pydict = {i: 'a' for i in range(cache_size)}
for i in range(22000):
    t = del_insert_pydict(pydict, i)
    if t > t_resize / 3:
        print(i)
# 0
# 10924
# 21848
# the dict is resized every 10924 items


# test lru_cache
@functools.lru_cache(maxsize=cache_size)
def who_am_i(a):
    return a

def t_who_am_i(a):
    s_time = time.time()
    who_am_i(a)
    e_time = time.time()
    return e_time - s_time

# populate the cache
for i in range(cache_size):
    who_am_i(i)

print('\nthe lru_cache is resized after caching the following number of items:')
for i in range(22000):
    t = t_who_am_i(i + cache_size)
    if t > t_resize / 3:
        print(i)
# 0
# 10924
# 21848
# the dict for the lru_cache is resized every 10924 items
```